search/features_layer_matcher.hpp

```diff
@@ -181,36 +183,74 @@ class FeaturesLayerMatcher
   template <typename TFn>
   void MatchPOIsWithStreets(FeaturesLayer const & child, FeaturesLayer const & parent, TFn && fn)
   {
+    BailIfCancelled(m_cancellable);
+
     ASSERT_EQUAL(child.m_type, Model::TYPE_POI, ());
     ASSERT_EQUAL(parent.m_type, Model::TYPE_STREET, ());
 
     auto const & pois = *child.m_sortedFeatures;
     auto const & streets = *parent.m_sortedFeatures;
 
-    // When the number of POIs is less than the number of STREETs,
-    // it's faster to check nearby streets for POIs.
-    if (pois.size() < streets.size())
+    vector<PointRectMatcher::PointIdPair> poiCenters;
+    poiCenters.reserve(pois.size());
+
+    for (size_t i = 0; i < pois.size(); ++i)
+    {
+      FeatureType poiFt;
+      GetByIndex(pois[i], poiFt);
+      poiCenters.emplace_back(feature::GetCenter(poiFt, FeatureType::WORST_GEOMETRY), i /* id */);
+    }
+
+    vector<PointRectMatcher::RectIdPair> streetRects;
+    streetRects.reserve(streets.size());
+
+    vector<ProjectionOnStreetCalculator> streetProjectors;
+    streetProjectors.reserve(streets.size());
+
+    for (size_t i = 0; i < streets.size(); ++i)
     {
-      for (uint32_t poiId : pois)
+      FeatureType streetFt;
+      GetByIndex(streets[i], streetFt);
+
+      streetFt.ParseGeometry(FeatureType::WORST_GEOMETRY);
+
+      m2::RectD inflationRect;
+      // Any point is good enough here, and feature::GetCenter would re-read the geometry.
+      if (streetFt.GetPointsCount() > 0)
       {
-        for (auto const & street : GetNearbyStreets(poiId))
-        {
-          if (street.m_distanceMeters > kStreetRadiusMeters)
-            break;
+        inflationRect = MercatorBounds::RectByCenterXYAndSizeInMeters(streetFt.GetPoint(0),
+                                                                      kStreetRadiusMeters);
+      }
 
-          uint32_t const streetId = street.m_id.m_index;
-          if (binary_search(streets.begin(), streets.end(), streetId))
-            fn(poiId, streetId);
-        }
+      for (size_t j = 0; j + 1 < streetFt.GetPointsCount(); ++j)
+      {
+        auto const & p1 = streetFt.GetPoint(j);
+        auto const & p2 = streetFt.GetPoint(j + 1);
+        m2::RectD rect(p1, p2);
+        rect.Inflate(inflationRect.SizeX(), inflationRect.SizeY());
```

Comment:
Looks like you need to Inflate by SizeX() * 0.5 and SizeY() * 0.5, right?

Reply:
In fact, even 0.25 is enough: RectByCenterXYAndSizeInMeters treats size as a radius, thus creating a 2S x 2S rectangle for the size parameter S.
After that, the Inflate() method adds strips of size dx both from the left and from the right.

done

Reply:
Disagree - we need 0.5. Imagine a case when the street is horizontal or vertical - we need the strip of size 2 * kStreetRadiusMeters around the street.

Reply:
Ah, it's because we need 2S x 2S in the end, as we treat a radius as, well, a radius. And not as the side length.

done